src/sys/windows/selector.rs

```diff
@@ -1,596 +1,543 @@
-        registry: &Registry,
-    ) -> io::Result<()> {
-        self.check_same_selector(registry)
-    }
-
-    /// Deregisters the handle provided from the `Poll` provided.
-    ///
-    /// This is intended to be used as part of `Evented::deregister` but note
-    /// that this function does not currently deregister the provided handle
-    /// from the `poll` specified. IOCP has a special binding for that which has
-    /// not yet been implemented. Instead this function should be used to assert
-    /// that the call to `deregister` happened on the same `Poll` that was
-    /// passed into to `register`.
-    ///
-    /// # Unsafety
-    ///
-    /// This function is unsafe for similar reasons to `register`. That is,
-    /// there may be pending I/O events and such which aren't handled correctly.
-    pub unsafe fn deregister_handle(
-        &self,
-        _handle: &dyn AsRawHandle,
-        registry: &Registry,
+        socket: &S,
+        token: Token,
+        interests: Interests,
     ) -> io::Result<()> {
-        self.check_same_selector(registry)
+        self.inner.reregister(socket, token, interests)
     }
 
-    /// Same as `deregister_handle`, but for sockets.
-    pub unsafe fn deregister_socket(
-        &self,
-        _socket: &dyn AsRawSocket,
-        registry: &Registry,
-    ) -> io::Result<()> {
-        self.check_same_selector(registry)
+    pub fn deregister<S: GenericSocket>(&self, socket: &S) -> io::Result<()> {
+        self.inner.deregister(socket)
     }
 
-    fn check_same_selector(&self, registry: &Registry) -> io::Result<()> {
-        let selector = poll::selector(registry);
-        match self.selector.borrow() {
-            Some(prev) if prev.identical(&selector.inner) => Ok(()),
-            Some(_) | None => Err(other("socket already registered")),
-        }
+    pub fn id(&self) -> usize {
+        self.id
     }
-}
 
-impl fmt::Debug for Binding {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Binding").finish()
+    pub(super) fn clone_inner(&self) -> Arc<SelectorInner> {
+        self.inner.clone()
     }
 }
 
-/// Helper struct used for TCP and UDP which bundles a `binding` with a
-/// `SetReadiness` handle.
-pub struct ReadyBinding {
-    binding: Binding,
-    readiness: Option<SetReadiness>,
+#[derive(Debug)]
+pub struct SelectorInner {
+    cp: CompletionPort,
+    active_poll_count: AtomicUsize,
+    update_queue: Mutex<VecDeque<Arc<Mutex<SockState>>>>,
+    afd_group: Mutex<Vec<Arc<Afd>>>,
 }
 
-impl ReadyBinding {
-    /// Creates a new blank binding ready to be inserted into an I/O object.
-    ///
-    /// Won't actually do anything until associated with an `Selector` loop.
-    pub fn new() -> ReadyBinding {
-        ReadyBinding {
-            binding: Binding::new(),
-            readiness: None,
-        }
+impl SelectorInner {
+    pub fn new() -> io::Result<SelectorInner> {
+        CompletionPort::new(0).map(|cp| SelectorInner {
```

Comment:
Is there a way to share the completion port with non-socket users that can't use the AFD_POLL.
In other words, how will mio-named-pipes work?

Reply:
In this scheme, selector cannot share with others its completion port, since the overlapped pointer is used for Arc::from_raw and NULL for wake

Reply:
So basically we need a more general solution for this issue. After all, not everything is a socket.
I think that mio needs to provide a way for non-socket I/O to use the same completion port (like named pipes, but also [asynchronous DNS resolution](https://docs.microsoft.com/en-us/windows/win32/api/ws2tcpip/nf-ws2tcpip-getaddrinfoexw)).

cc @carllerche 
I stubbed out some not-quite-complete idea I have for this in https://github.com/piscisaureus/iocp-traits

Reply:
Then Selector will need to have a workaround for detecting inappropriate overlapped to skip. Any idea?

Reply:
@piscisaureus your strategy is fine to add. Maybe we can tackle it as a follow up after this PR lands?

Reply:
I agree this can land first.